File: src/ui.py
"""Rich CLI interface for the Aristotle Agent."""

import random
import threading
import time

from rich.console import Console
from rich.live import Live
from rich.markdown import Markdown
from rich.panel import Panel
from rich.rule import Rule
from rich.spinner import Spinner
from rich.table import Table
from rich.text import Text
from rich.theme import Theme

from src.inference import AristotleAgent, RetrievedChunk

# ── Theme ──────────────────────────────────────────────────────────────────
custom_theme = Theme({
    "aristotle": "bold gold1",
    "citation": "cyan",
    "debug": "magenta",
    "user": "bold green",
    "info": "dim",
})

console = Console(theme=custom_theme)

_THINKING_VERBS = [
    "Contemplating",
    "Reasoning",
    "Deliberating",
    "Reflecting",
    "Considering",
    "Examining",
    "Analyzing",
    "Pondering",
    "Meditating",
    "Inquiring",
    "Discerning",
    "Syllogizing",
]

_LOADING_PHASES = [
    "Unrolling the scrolls",
    "Recalling the Categories",
    "Reviewing the Ethics",
    "Ordering first principles",
    "Warming up the Lyceum",
    "Distinguishing genus from species",
    "Preparing the dialectic",
    "Recollecting the virtues",
    "Tuning the lyre of reason",
]

ASCII_LOGO = r"""
 █████╗ ██████╗ ██╗███████╗████████╗ ██████╗ ████████╗██╗     ███████╗       ██████╗██╗     ██╗
██╔══██╗██╔══██╗██║██╔════╝╚══██╔══╝██╔═══██╗╚══██╔══╝██║     ██╔════╝      ██╔════╝██║     ██║
███████║██████╔╝██║███████╗   ██║   ██║   ██║   ██║   ██║     █████╗  █████╗██║     ██║     ██║
██╔══██║██╔══██╗██║╚════██║   ██║   ██║   ██║   ██║   ██║     ██╔══╝  ╚════╝██║     ██║     ██║
██║  ██║██║  ██║██║███████║   ██║   ╚██████╔╝   ██║   ███████╗███████╗      ╚██████╗███████╗██║
╚═╝  ╚═╝╚═╝  ╚═╝╚═╝╚══════╝   ╚═╝    ╚═════╝    ╚═╝   ╚══════╝╚══════╝       ╚═════╝╚══════╝╚═╝""".lstrip("\n")


# ── Display Functions ──────────────────────────────────────────────────────

def print_welcome() -> None:
    """Print styled welcome banner."""
    banner = (
        f"[aristotle]{ASCII_LOGO}[/aristotle]\n\n"
        "[info]An agent grounded in the Nicomachean Ethics[/info]\n\n"
        "Ask a question, or type [bold]/help[/bold] for commands."
    )
    console.print(Panel(banner, border_style="gold1", padding=(1, 2)))


def print_debug_retrieval(chunks: list[RetrievedChunk]) -> None:
    """Display retrieved chunks as a Rich table."""
    table = Table(title="Retrieved Passages", border_style="debug")
    table.add_column("#", style="bold", width=3)
    table.add_column("Score", style="debug", width=7)
    table.add_column("Source", style="citation", width=18)
    table.add_column("Tags", width=30)
    table.add_column("Preview", ratio=1)

    for i, chunk in enumerate(chunks, 1):
        source = f"Book {chunk.book}, Ch. {chunk.chapter}"
        tags = ", ".join(chunk.tags) if chunk.tags else ""
        preview = chunk.text[:120].replace("\n", " ") + "..."
        table.add_row(
            str(i),
            f"{chunk.score:.3f}",
            source,
            tags,
            preview,
        )

    console.print(table)


def print_debug_prompt(messages: list[dict]) -> None:
    """Display the assembled prompt (user message only, skip system)."""
    prompt_text = ""
    for msg in messages:
        if msg["role"] == "system":
            continue
        role = msg["role"].upper()
        prompt_text += f"--- {role} ---\n{msg['content']}\n\n"

    console.print(Panel(
        prompt_text,
        title="[debug]Assembled Prompt[/debug]",
        border_style="magenta",
        expand=True,
    ))


def print_help() -> None:
    """Print available commands."""
    help_text = (
        "[bold]Commands:[/bold]\n"
        "  [bold]/debug[/bold]   — Toggle debug mode (show retrieval + prompt)\n"
        "  [bold]/help[/bold]    — Show this help message\n"
        "  [bold]exit[/bold]     — Quit the program\n"
        "  [bold]quit[/bold]     — Quit the program"
    )
    console.print(Panel(help_text, border_style="info"))


def get_user_input() -> str:
    """Get user input with styled prompt."""
    try:
        console.print()
        console.print(Rule(style="dim"))                        # top rule
        user_input = console.input("[user]You:[/user] ").strip()
        console.print(Rule(style="dim"))                        # bottom rule (after Enter)
        if user_input:
            console.print()                                     # breathing room before response
        return user_input
    except (EOFError, KeyboardInterrupt):
        return "exit"


def stream_response(agent: AristotleAgent, query: str) -> None:
    """Run the query pipeline and stream the response."""
    result: dict = {}

    def _run_ask() -> None:
        result["value"] = agent.ask(query, stream=True)

    worker = threading.Thread(target=_run_ask, daemon=True)

    # Shuffle verbs so each query gets a unique sequence
    verbs = random.sample(_THINKING_VERBS, len(_THINKING_VERBS))
    verb_idx = 0
    spinner = Spinner("dots", text=f"[info] {verbs[0]}...[/info]", style="aristotle")

    with Live(spinner, console=console, transient=True, refresh_per_second=12):
        worker.start()
        while worker.is_alive():
            worker.join(timeout=0.8)
            verb_idx = (verb_idx + 1) % len(verbs)
            spinner.update(text=f"[info] {verbs[verb_idx]}...[/info]")

    response, token_gen = result["value"]

    if response.debug:
        console.print()
        print_debug_retrieval(response.debug.chunks)
        console.print()
        print_debug_prompt(response.debug.prompt)
        console.print()

    # Stream response
    console.print("[aristotle]Aristotle:[/aristotle]")
    for token in token_gen:
        print(token, end="", flush=True)
    print()  # newline after streaming


# ── Main Loop ──────────────────────────────────────────────────────────────

def run_cli(agent: AristotleAgent) -> None:
    """Main interactive loop."""
    print_welcome()

    # Eagerly load embeddings + ChromaDB so first query is fast
    def _load() -> None:
        agent._ensure_retriever()

    worker = threading.Thread(target=_load, daemon=True)
    phases = random.sample(_LOADING_PHASES, len(_LOADING_PHASES))
    phase_idx = 0
    spinner = Spinner("dots", text=f"[info] {phases[0]}...[/info]", style="aristotle")

    with Live(spinner, console=console, transient=True, refresh_per_second=12):
        worker.start()
        while worker.is_alive():
            worker.join(timeout=2.5)
            phase_idx = (phase_idx + 1) % len(phases)
            spinner.update(text=f"[info] {phases[phase_idx]}...[/info]")

    while True:
        query = get_user_input()

        if not query:
            continue

        if query.lower() in ("exit", "quit"):
            console.print("[info]Farewell.[/info]")
            break

        if query.lower() == "/help":
            print_help()
            continue

        if query.lower() == "/debug":
            agent.debug = not agent.debug
            state = "ON" if agent.debug else "OFF"
            console.print(f"[debug]Debug mode: {state}[/debug]")
            continue

        try:
            stream_response(agent, query)
        except Exception as e:
            console.print(f"[bold red]Error:[/bold red] {e}")

File: src/ui.ts
/**
 * CLI interface — banner, spinners, streaming output, input loop.
 */

import { createInterface } from "node:readline/promises";
import { stdin, stdout } from "node:process";
import chalk from "chalk";
import ora from "ora";
import Table from "cli-table3";
import { AristotleAgent, type DebugInfo } from "./agent.js";
import type { RetrievedChunk } from "./retriever.js";

// ── Constants ────────────────────────────────────────────────────────────

const gold = chalk.hex("#FFD700");
const goldBold = chalk.hex("#FFD700").bold;
const dim = chalk.dim;
const green = chalk.green.bold;
const magenta = chalk.magenta;

const THINKING_VERBS = [
  "Contemplating",
  "Reasoning",
  "Deliberating",
  "Reflecting",
  "Considering",
  "Examining",
  "Analyzing",
  "Pondering",
  "Meditating",
  "Inquiring",
  "Discerning",
  "Syllogizing",
];

const LOADING_PHASES = [
  "Unrolling the scrolls",
  "Recalling the Categories",
  "Reviewing the Ethics",
  "Ordering first principles",
  "Warming up the Lyceum",
  "Distinguishing genus from species",
  "Preparing the dialectic",
  "Recollecting the virtues",
  "Tuning the lyre of reason",
  "Consulting the Metaphysics",
  "Reviewing the Politics",
  "Studying the Rhetoric",
  "Contemplating the Physics",
];

const ASCII_LOGO = `
 ${goldBold("█████╗ ██████╗ ██╗███████╗████████╗ ██████╗ ████████╗██╗     ███████╗       ██████╗██╗     ██╗")}
${goldBold("██╔══██╗██╔══██╗██║██╔════╝╚══██╔══╝██╔═══██╗╚══██╔══╝██║     ██╔════╝      ██╔════╝██║     ██║")}
${goldBold("███████║██████╔╝██║███████╗   ██║   ██║   ██║   ██║   ██║     █████╗  █████╗██║     ██║     ██║")}
${goldBold("██╔══██║██╔══██╗██║╚════██║   ██║   ██║   ██║   ██║   ██║     ██╔══╝  ╚════╝██║     ██║     ██║")}
${goldBold("██║  ██║██║  ██║██║███████║   ██║   ╚██████╔╝   ██║   ███████╗███████╗      ╚██████╗███████╗██║")}
${goldBold("╚═╝  ╚═╝╚═╝  ╚═╝╚═╝╚══════╝   ╚═╝    ╚═════╝    ╚═╝   ╚══════╝╚══════╝       ╚═════╝╚══════╝╚═╝")}
`;

const GLITCH_CHARS = "αβγδεζηθικλμνξπρστφχψω";
const GLITCH_LEN = 5;
const CHAR_DELAY = 8; // ms — ~125 chars/sec

// ── Helpers ──────────────────────────────────────────────────────────────

function shuffle<T>(arr: T[]): T[] {
  const a = [...arr];
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function randomGlitchChar(): string {
  return GLITCH_CHARS[Math.floor(Math.random() * GLITCH_CHARS.length)];
}

function stripAnsi(str: string): string {
  // eslint-disable-next-line no-control-regex
  return str.replace(/\x1b\[[0-9;]*m/g, "");
}

// ── Display Functions ────────────────────────────────────────────────────

function printWelcome(): void {
  const contentLines = [
    ...ASCII_LOGO.split("\n").filter(Boolean),
    "",
    dim("  An agent grounded in the complete works of Aristotle"),
    "",
    `  Ask a question, or type ${chalk.bold("/help")} for commands.`,
  ];

  const hPad = 2;
  const vPad = 1;
  const maxWidth = Math.max(...contentLines.map((l) => stripAnsi(l).length));
  const innerWidth = maxWidth + hPad * 2;

  const emptyRow = gold("│") + " ".repeat(innerWidth) + gold("│");

  console.log();
  console.log(gold(`╭${"─".repeat(innerWidth)}╮`));
  for (let i = 0; i < vPad; i++) console.log(emptyRow);
  for (const line of contentLines) {
    const visualLen = stripAnsi(line).length;
    const rightPad = maxWidth - visualLen + hPad;
    console.log(
      gold("│") + " ".repeat(hPad) + line + " ".repeat(rightPad) + gold("│"),
    );
  }
  for (let i = 0; i < vPad; i++) console.log(emptyRow);
  console.log(gold(`╰${"─".repeat(innerWidth)}╯`));
  console.log();
}

function printDebugRetrieval(chunks: RetrievedChunk[]): void {
  const table = new Table({
    head: [
      magenta("#"),
      magenta("Score"),
      chalk.cyan("Source"),
      "Tags",
      "Preview",
    ],
    colWidths: [4, 9, 20, 32, 50],
    wordWrap: true,
  });

  for (let i = 0; i < chunks.length; i++) {
    const c = chunks[i];
    const source = `${c.work}\nBk ${c.book}, Pt ${c.chapter}`;
    const tags = c.tags.join(", ");
    const preview = c.text.slice(0, 120).replace(/\n/g, " ") + "...";
    table.push([
      String(i + 1),
      magenta(c.score.toFixed(3)),
      chalk.cyan(source),
      tags,
      preview,
    ]);
  }

  console.log();
  console.log(magenta("  Retrieved Passages"));
  console.log(table.toString());
}

function printDebugPrompt(
  prompt: Array<{ role: string; content: string }>,
): void {
  console.log();
  console.log(magenta("  Assembled Prompt"));
  for (const msg of prompt) {
    if (msg.role === "system") continue;
    console.log(magenta(`  --- ${msg.role.toUpperCase()} ---`));
    console.log(msg.content);
  }
  console.log();
}

function printHelp(): void {
  console.log();
  console.log(chalk.bold("  Commands:"));
  console.log(`  ${chalk.bold("/debug")}   — Toggle debug mode (show retrieval + prompt)`);
  console.log(`  ${chalk.bold("/help")}    — Show this help message`);
  console.log(`  ${chalk.bold("exit")}     — Quit the program`);
  console.log(`  ${chalk.bold("quit")}     — Quit the program`);
  console.log();
}

// ── Spinner with cycling text ────────────────────────────────────────────

function cyclingSpinner(
  phrases: string[],
  intervalMs: number,
): { spinner: ReturnType<typeof ora>; stop: () => void } {
  const shuffled = shuffle(phrases);
  let idx = 0;
  const spinner = ora({
    text: dim(` ${shuffled[0]}...`),
    color: "yellow",
  }).start();

  const timer = setInterval(() => {
    idx = (idx + 1) % shuffled.length;
    spinner.text = dim(` ${shuffled[idx]}...`);
  }, intervalMs);

  return {
    spinner,
    stop: () => {
      clearInterval(timer);
      spinner.stop();
    },
  };
}

// ── Greek glitch-trail typewriter ────────────────────────────────────────

class GlitchTypewriter {
  private col = 0;
  private terminalWidth: number;
  private wordBuf: string[] = [];
  private glitchBuf: string[] = [];

  constructor() {
    this.terminalWidth = stdout.columns || 80;
  }

  private showGlitch(): void {
    const avail = this.terminalWidth - this.col;
    const n = Math.min(GLITCH_LEN, avail);
    if (n <= 0) {
      this.glitchBuf.length = 0;
      return;
    }
    if (this.glitchBuf.length > 0) {
      this.glitchBuf.shift();
      this.glitchBuf.push(randomGlitchChar());
    }
    while (this.glitchBuf.length < n) {
      this.glitchBuf.push(randomGlitchChar());
    }
    this.glitchBuf.length = n;
    const trail = this.glitchBuf.join("");
    stdout.write(`\x1b[s${trail}\x1b[u`);
  }

  private clearGlitch(): void {
    const n = this.glitchBuf.length;
    if (n <= 0) return;
    stdout.write("\x1b[K");
    this.glitchBuf.length = 0;
  }

  private async flushWord(): Promise<void> {
    const word = this.wordBuf.join("");
    this.wordBuf.length = 0;
    if (!word) return;

    const needed = (this.col > 0 ? 1 : 0) + word.length;
    if (this.col + needed > this.terminalWidth && this.col > 0) {
      this.clearGlitch();
      stdout.write("\n");
      this.col = 0;
    }
    if (this.col > 0) {
      stdout.write(" ");
      await sleep(CHAR_DELAY);
      this.col++;
      this.showGlitch();
    }
    for (const c of word) {
      stdout.write(c);
      await sleep(CHAR_DELAY);
      this.col++;
      this.showGlitch();
    }
  }

  async write(tokenGen: AsyncGenerator<string>): Promise<void> {
    // Hide cursor during typewriter
    stdout.write("\x1b[?25l");

    let pendingNewlines = 0;

    for await (const token of tokenGen) {
      for (const ch of token) {
        if (ch === "\n") {
          await this.flushWord();
          this.clearGlitch();
          pendingNewlines++;
        } else if (ch === " ") {
          if (pendingNewlines > 0) {
            for (let i = 0; i < Math.min(pendingNewlines, 1); i++) {
              stdout.write("\n");
            }
            this.col = 0;
            pendingNewlines = 0;
          }
          await this.flushWord();
        } else {
          if (pendingNewlines > 0) {
            for (let i = 0; i < Math.min(pendingNewlines, 1); i++) {
              stdout.write("\n");
            }
            this.col = 0;
            pendingNewlines = 0;
          }
          this.wordBuf.push(ch);
        }
      }
    }

    await this.flushWord();
    this.clearGlitch();
    // Restore cursor and add scroll padding
    stdout.write("\x1b[?25h");
    stdout.write("\n");
  }
}

// ── Stream response ──────────────────────────────────────────────────────

async function streamResponse(
  agent: AristotleAgent,
  query: string,
): Promise<void> {
  // Show thinking spinner while retrieving + building prompt
  const thinking = cyclingSpinner(THINKING_VERBS, 800);

  let response: Awaited<ReturnType<typeof agent.ask>>;
  try {
    response = await agent.ask(query);
  } finally {
    thinking.stop();
  }

  const [agentResponse, tokenGen] = response;

  if (agentResponse.debug) {
    printDebugRetrieval(agentResponse.debug.chunks);
    printDebugPrompt(agentResponse.debug.prompt);
  }

  // Print label
  stdout.write(goldBold("Aristotle: ") + "\n");

  // Wait for first token with spinner, then stop spinner before typewriter starts
  const firstTokenSpinner = cyclingSpinner(THINKING_VERBS, 800);
  const firstResult = await tokenGen.next();
  firstTokenSpinner.stop();

  if (firstResult.done) return;

  async function* prependFirst(): AsyncGenerator<string> {
    yield firstResult.value;
    yield* tokenGen;
  }

  const typewriter = new GlitchTypewriter();
  await typewriter.write(prependFirst());
}

// ── Main CLI loop ────────────────────────────────────────────────────────

export async function runCli(agent: AristotleAgent): Promise<void> {
  printWelcome();

  // Eagerly load embeddings + download model with spinning status
  const loading = cyclingSpinner(LOADING_PHASES, 2500);
  try {
    await agent.ensureRetriever();
  } finally {
    loading.stop();
  }

  while (true) {
    console.log();

    const rl = createInterface({ input: stdin, output: stdout });
    let query: string;
    try {
      query = (await rl.question(green("You: "))).trim();
    } catch {
      rl.close();
      break;
    }
    rl.close();

    console.log();

    if (!query) continue;

    const lower = query.toLowerCase();
    if (lower === "exit" || lower === "quit") {
      console.log(dim("Farewell."));
      break;
    }

    if (lower === "/help") {
      printHelp();
      continue;
    }

    if (lower === "/debug") {
      agent.debug = !agent.debug;
      const state = agent.debug ? "ON" : "OFF";
      console.log(magenta(`Debug mode: ${state}`));
      continue;
    }

    try {
      await streamResponse(agent, query);
    } catch (err) {
      console.error(chalk.red(`Error: ${err instanceof Error ? err.message : err}`));
    }
  }
}
